README.md

```diff
@@ -1,41 +1,102 @@
+The client is prompted to supply a nickname on starting the program. Once supplied, the client sends the server a connect message with this nickname. When the server receives a connect message, the server sends the client an acknowledgement message and adds the client to the clients list. The client keeps resending the server connect messages (waiting a bit between resends) until getting an acknowledgement from the server.
+
+### Chatting
+
+The server stores two (main) pieces of state information: a list of clients connected, and the most recent 45 characters of chatting. The list of clients connected is a hash, with information about the 'color' (an integer) the server has assigned the client, the time of the last message received from the client, the client's user supplied nickname, and the client's address information (host and port). The data for the most recent 45 characters of chatting includes info for each character about what the character is, and the 'color' (assigned integer) of the client it came from.
+
+When the client presses a (alphanumeric, space, or punctuation) key, the client sends a chat message to the server.
+
+When the server receives a [chat message from a client](#sent-by-client), the server checks its timestamp and compares it to the time of the last message received from that client. If it's older than the last message received, the message is ignored. Assuming it's a new message, the server sends all clients a [chat message](#sent-by-server) with data about the nicknames of clients in the client list and who is the 'speaker', and data representing the state of the last 45 characters of chat text as described above.
+
+When the client receives such a chat message from the server, the client checks the timestamp to make sure it's newer than the last chat message received, and if so the client overwrites the chat names list and chat text in the terminal to reflect the updated state.
+
+### Disconnecting
+When a client quits properly, that is by typing the quit command `\`, the client sends the server a quit message and the program exits. When the server receives a quit message from a client, the server removes the client from the stored clients list, and adds the 'color' (integer) associated with the client back to the list of available colors.
+
+
+## Message formats
+
+### Sent by client
 
 connect:
-{'type' => 'connect', 'name' => (String), 'time' => (timestamp String)}
+`{'type' => 'connect', 'name' => (String), 'time' => (timestamp String)}`
 
 chat:
-{'type' => 'chat', 'body' => (String), 'time' => (timestamp String)}
+`{'type' => 'chat', 'body' => (String), 'time' => (timestamp String)}`
 
 quit:
-{'type' => 'quit', 'time' => (timestamp String)}
+`{'type' => 'quit', 'time' => (timestamp String)}`
+
 
-exists:
-{'type' => 'hi', 'time' => (timestamp String)}
-- not yet implemented
+### Sent by server
 
-Sent by server:
+chat: `{'type' => 'chat', 'body' => (Array), 'names' => (Array) 'time' => (timestamp String)}`
```

Comment:
I'm interested to see a concrete example of a `body`. Given two users, it might look like this?
```
user 0 says "foobar "
user 1 says "bazbam"
results in:
f0o0o0b1a1z1b0a0r0 0b1a1m1
```